README.md
A Dart library for implementing the encoding and decoding of the Bencode format.

All codes come from [bencode.js][bencode.js], include example and test codes , I just transfer them to Dart code.

## Install

In your flutter or dart project add the dependency:

```
dependencies:
  b_encode_decode: ^1.0.3
```

## Usage

A simple usage example:

### Encode

Input parameter can be a String, Number, List, or Map. It will return a encoding bytes list ( `Uint8List` ).

```dart
import 'package:b_encode_decode/b_encode_decode.dart' as Bencode;

main() {
  Bencode.encode("string")         // => "6:string"
  Bencode.encode(123)              // => "i123e"
  Bencode.encode(["str", 123])     // => "l3:stri123ee"
  Bencode.encode({ "key": "value" }) // => "d3:key5:valuee"
}
```

### Decode

Input should be bytes list or String.

```dart
import 'package:b_encode_decode/b_encode_decode.dart' as Bencode;

main() {
  var map = Bencode.decode(Uint8List.fromList('d3:key5:valuee'.codeUnits); // => { key: "value" } , the string value is bytes array

  print(map);
```

## Features and bugs

Please file feature requests and bugs at the [issue tracker][tracker].

[tracker]: https://github.com/moham96/b_encode_decode/issues
[bencode.js]: https://github.com/benjreinhart/bencode-js


CHANGELOG.md
## 1.0.0

- Initial version

## 1.0.1

- Change Readme and add example

## 1.0.2

- Fix string encoding error

## 1.0.3

- Migrate to nullsafety
- update dependencies
- some linting and code style enhancments

## 1.0.4
- rename example

test/bencode_test.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:bencode_dart/bencode_dart.dart';
import 'package:test/test.dart';

/// All tests come from https://github.com/benjreinhart/bencode-js/blob/master/test
///
/// I just transfer JS to Dart
void main() {
  group('Tests for decoding - ', () {
    test('It with a basic value returns the string', () {
      var code = decode(stringToBytes('10:helloworld'));
      var str = (code is Uint8List) ? bytesToString(code) : code;
      assert(str == 'helloworld');
    });

    test('It with a colon in them returns the correct string', () {
      var code = decode(stringToBytes('12:0.0.0.0:3000'));
      var str = (code is Uint8List) ? bytesToString(code) : code;
      assert(str == '0.0.0.0:3000');
    });

    test('It is the integer between i and e', () {
      var code = decode(stringToBytes('i42e'));
      assert(code == 42);
    });

    test('It allows negative numbers', () {
      var code = decode(stringToBytes('i-42e'));
      assert(code == -42);
    });

    test('It allows zeros', () {
      var code = decode(stringToBytes('i0e'));
      assert(code == 0);
    });

    test('It creates a list with strings and integers', () {
      var re = decode(stringToBytes('l5:helloi42ee'));
      var c = ['hello', 42];
      assert(myEqauls(re, c));
    });

    test('It creates a list with nested lists of strings and integers', () {
      var re = decode(stringToBytes('l5:helloi42eli-1ei0ei1ei2ei3e4:fouree'));
      var c = [
        'hello',
        42,
        [-1, 0, 1, 2, 3, 'four']
      ];
      assert(myEqauls(re, c));
    });

    test('It has no problem with multiple empty lists or objects', () {
      var re = decode(stringToBytes('lllleeee'));
      var c = [
        [
          [[]]
        ]
      ];
      assert(myEqauls(re, c));

      re = decode(stringToBytes('llelelelleee'));
      var c1 = [
        [],
        [],
        [],
        [[]]
      ];
      assert(myEqauls(re, c1));

      re = decode(stringToBytes('ldededee'));
      var c2 = [{}, {}, {}];
      assert(myEqauls(re, c2));
    });

    test('It creates an object with strings and integers', () {
      var re = decode(stringToBytes('d3:agei100e4:name8:the dudee'));
      var c = {'age': 100, 'name': 'the dude'};
      assert(myEqauls(re, c));
    });

    test('It creates an object with nested objects of strings and integers',
        () {
      var re = decode(stringToBytes(
          'd3:agei100e4:infod5:email13:dude@dude.com6:numberi2488081446ee4:name8:the dudee'));
      var c = {
        'age': 100,
        'info': {'email': 'dude@dude.com', 'number': 2488081446},
        'name': 'the dude'
      };
      assert(myEqauls(re, c));
    });

    test('It has no problem with an empty object', () {
      var re = decode(stringToBytes('de'));
      var c = {};
      assert(myEqauls(re, c));
    });

    test('It creates an object with a list of objects', () {
      var re = decode(stringToBytes(
          'd9:locationsld7:address10:484 streeted7:address10:828 streeteee'));
      var c = {
        'locations': [
          {'address': '484 street'},
          {'address': '828 street'}
        ]
      };
      assert(myEqauls(re, c));
    });

    test('It has no problem when there are multiple "e"s in a row', () {
      var re = decode(stringToBytes('lld9:favoritesleeei500ee'));
      var c = [
        [
          {'favorites': []}
        ],
        500
      ];
      assert(myEqauls(re, c));
    });
  });

  // encoding:
  group('Tests for encoding - ', () {
    test('It encodes a string as <lenOfString>:<string>', () {
      var re = encode('omg hay thurrr');
      var c = '14:omg hay thurrr';
      assert(c == bytesToString(re));
    });

    test('It encodes integers as i<integer>e', () {
      var re = encode(2234);
      var c = 'i2234e';
      assert(c == bytesToString(re));
    });

    test('It encodes negative integers', () {
      var re = bytesToString(encode(-2234));
      var c = 'i-2234e';
      assert(c == re);
    });

    test('It encodes large-ish integers', () {
      var re = encode(2222222222);
      var c = 'i2222222222e';
      assert(c == bytesToString(re));
    });

    test('It encodes a list as l<list items>e', () {
      var re = encode(['a string', 23]);
      var c = 'l8:a stringi23ee';
      assert(c == bytesToString(re));
    });

    test('It encodes empty lists', () {
      var re = encode([]);
      var c = 'le';
      assert(c == bytesToString(re));
    });

    test('It encodes nested lists', () {
      var re = encode([
        ['james', 'john'],
        [
          ['jordin', 12]
        ]
      ]);
      var c = 'll5:james4:johnell6:jordini12eeee';
      assert(c == bytesToString(re));
    });

    test('It encodes an object as d<key><value>e where keys are sorted', () {
      var re = encode({'name': 'ben', 'age': 23});
      var c = 'd3:agei23e4:name3:bene';
      assert(c == bytesToString(re));
    });

    test('It encodes empty objects', () {
      var re = encode({});
      var c = 'de';
      assert(c == bytesToString(re));
    });

    test('It nested objects and lists', () {
      var re = encode({
        'people': [
          {'name': 'j', 'age': 20}
        ]
      });
      var c = 'd6:peopleld3:agei20e4:name1:jeee';
      assert(c == bytesToString(re));
    });
  });
}

Uint8List stringToBytes(str) {
  return utf8.encode(str);
}

int bytesToInt(bytes) {
  return bytes[0];
}

bool myEqauls(s, t) {
  if (t is List) {
    if (s is! List) return false;
    if (s.length != t.length) return false;
    for (var i = 0; i < s.length; i++) {
      if (!myEqauls(s[i], t[i])) return false;
    }
    return true;
  }
  if (t is Map) {
    if (s is! Map) return false;
    if (s.length != t.length) return false;
    var keys = t.keys.toList();
    var keys2 = s.keys.toList();
    if (!myEqauls(keys, keys2)) return false;
    for (var key in keys) {
      if (!myEqauls(s[key], t[key])) return false;
    }
    return true;
  }
  if (t is String) {
    if (s is! String) return bytesToString(s) == t;
    return s == t;
  }
  if (t is num) return s == t;
  return false;
}

String bytesToString(bytes) {
  return utf8.decode(bytes);
}


test/bencode_dart_test.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:b_encode_decode/b_encode_decode.dart';
import 'package:test/test.dart';

/// All tests come from https://github.com/benjreinhart/bencode-js/blob/master/test
///
/// I just transfer JS to Dart
void main() {
  group('Tests for decoding - ', () {
    test('It with a basic value returns the string', () {
      var code = decode(stringToBytes('10:helloworld'));
      var str = (code is Uint8List) ? bytesToString(code) : code;
      assert(str == 'helloworld');
    });

    test('It with a colon in them returns the correct string', () {
      var code = decode(stringToBytes('12:0.0.0.0:3000'));
      var str = (code is Uint8List) ? bytesToString(code) : code;
      assert(str == '0.0.0.0:3000');
    });

    test('It is the integer between i and e', () {
      var code = decode(stringToBytes('i42e'));
      assert(code == 42);
    });

    test('It allows negative numbers', () {
      var code = decode(stringToBytes('i-42e'));
      assert(code == -42);
    });

    test('It allows zeros', () {
      var code = decode(stringToBytes('i0e'));
      assert(code == 0);
    });

    test('It creates a list with strings and integers', () {
      var re = decode(stringToBytes('l5:helloi42ee'));
      var c = ['hello', 42];
      assert(myEqauls(re, c));
    });

    test('It creates a list with nested lists of strings and integers', () {
      var re = decode(stringToBytes('l5:helloi42eli-1ei0ei1ei2ei3e4:fouree'));
      var c = [
        'hello',
        42,
        [-1, 0, 1, 2, 3, 'four']
      ];
      assert(myEqauls(re, c));
    });

    test('It has no problem with multiple empty lists or objects', () {
      var re = decode(stringToBytes('lllleeee'));
      var c = [
        [
          [[]]
        ]
      ];
      assert(myEqauls(re, c));

      re = decode(stringToBytes('llelelelleee'));
      var c1 = [
        [],
        [],
        [],
        [[]]
      ];
      assert(myEqauls(re, c1));

      re = decode(stringToBytes('ldededee'));
      var c2 = [{}, {}, {}];
      assert(myEqauls(re, c2));
    });

    test('It creates an object with strings and integers', () {
      var re = decode(stringToBytes('d3:agei100e4:name8:the dudee'));
      var c = {'age': 100, 'name': 'the dude'};
      assert(myEqauls(re, c));
    });

    test('It creates an object with nested objects of strings and integers',
        () {
      var re = decode(stringToBytes(
          'd3:agei100e4:infod5:email13:dude@dude.com6:numberi2488081446ee4:name8:the dudee'));
      var c = {
        'age': 100,
        'info': {'email': 'dude@dude.com', 'number': 2488081446},
        'name': 'the dude'
      };
      assert(myEqauls(re, c));
    });

    test('It has no problem with an empty object', () {
      var re = decode(stringToBytes('de'));
      var c = {};
      assert(myEqauls(re, c));
    });

    test('It creates an object with a list of objects', () {
      var re = decode(stringToBytes(
          'd9:locationsld7:address10:484 streeted7:address10:828 streeteee'));
      var c = {
        'locations': [
          {'address': '484 street'},
          {'address': '828 street'}
        ]
      };
      assert(myEqauls(re, c));
    });

    test('It has no problem when there are multiple "e"s in a row', () {
      var re = decode(stringToBytes('lld9:favoritesleeei500ee'));
      var c = [
        [
          {'favorites': []}
        ],
        500
      ];
      assert(myEqauls(re, c));
    });
  });

  // encoding:
  group('Tests for encoding - ', () {
    test('It encodes a string as <lenOfString>:<string>', () {
      var re = encode('omg hay thurrr');
      var c = '14:omg hay thurrr';
      assert(c == bytesToString(re));
    });

    test('It encodes integers as i<integer>e', () {
      var re = encode(2234);
      var c = 'i2234e';
      assert(c == bytesToString(re));
    });

    test('It encodes negative integers', () {
      var re = bytesToString(encode(-2234));
      var c = 'i-2234e';
      assert(c == re);
    });

    test('It encodes large-ish integers', () {
      var re = encode(2222222222);
      var c = 'i2222222222e';
      assert(c == bytesToString(re));
    });
    test('It throws an exeption when passed -0', () {
      expect(() => encode(-0), throwsException);
    });
    test('It encodes integers with leading zeros', () {
      var re = encode(00002);
      var c = 'i2e';
      assert(c == bytesToString(re));
    });
    test('It encodes a list as l<list items>e', () {
      var re = encode(['a string', 23]);
      var c = 'l8:a stringi23ee';
      assert(c == bytesToString(re));
    });

    test('It encodes empty lists', () {
      var re = encode([]);
      var c = 'le';
      assert(c == bytesToString(re));
    });

    test('It encodes nested lists', () {
      var re = encode([
        ['james', 'john'],
        [
          ['jordin', 12]
        ]
      ]);
      var c = 'll5:james4:johnell6:jordini12eeee';
      assert(c == bytesToString(re));
    });

    test('It encodes an object as d<key><value>e where keys are sorted', () {
      var re = encode({'name': 'ben', 'age': 23});
      var c = 'd3:agei23e4:name3:bene';
      assert(c == bytesToString(re));
    });

    test('It encodes empty objects', () {
      var re = encode({});
      var c = 'de';
      assert(c == bytesToString(re));
    });

    test('It nested objects and lists', () {
      var re = encode({
        'people': [
          {'name': 'j', 'age': 20}
        ]
      });
      var c = 'd6:peopleld3:agei20e4:name1:jeee';
      assert(c == bytesToString(re));
    });
  });
}

Uint8List stringToBytes(str) {
  return Uint8List.fromList(utf8.encode(str));
}

int bytesToInt(bytes) {
  return bytes[0];
}

bool myEqauls(s, t) {
  if (t is List) {
    if (s is! List) return false;
    if (s.length != t.length) return false;
    for (var i = 0; i < s.length; i++) {
      if (!myEqauls(s[i], t[i])) return false;
    }
    return true;
  }
  if (t is Map) {
    if (s is! Map) return false;
    if (s.length != t.length) return false;
    var keys = t.keys.toList();
    var keys2 = s.keys.toList();
    if (!myEqauls(keys, keys2)) return false;
    for (var key in keys) {
      if (!myEqauls(s[key], t[key])) return false;
    }
    return true;
  }
  if (t is String) {
    if (s is! String) return bytesToString(s) == t;
    return s == t;
  }
  if (t is num) return s == t;
  return false;
}

String bytesToString(bytes) {
  return utf8.decode(bytes);
}


lib/src/bencode_base.dart
import 'dart:convert';
import 'dart:typed_data';

///
/// Encode objects to bencoding format bytes;
///
/// This method comes from https://github.com/benjreinhart/bencode-js , just transfer JS to Dart
Uint8List encode(dynamic data, [dynamic buffer, int? offset]) {
  if (data == null) return Uint8List(0);
  return _Encode(data, buffer, offset).encoding();
}

class _Encode {
  int bytes = -1;
  final dynamic _data;
  final Uint8List? _buffer;
  final int? _offset;

  var buffE = Uint8List.fromList(utf8.encode('e')); // Buffer.from('e');
  var buffD = Uint8List.fromList(utf8.encode('d')); //Buffer.from('d');
  var buffL = Uint8List.fromList(utf8.encode('l')); //Buffer.from('l');

  _Encode(this._data, [this._buffer, this._offset]);

  Uint8List encoding() {
    var buffers = <Uint8List>[];
    Uint8List result;

    _encode(buffers, _data);
    result =
        Uint8List.fromList(buffers.fold<List<int>>([], (previousValue, buffer) {
      previousValue.addAll(buffer);
      return previousValue;
    }));
    bytes = result.length;

    if (_buffer != null) {
      (_buffer as Uint8List).insertAll(_offset ?? 0, result);
      return _buffer!;
    }

    return result;
  }

  void _encode(buffers, data) {
    if (data == null) {
      return;
    }

    if (data is Uint8List) {
      buffer(buffers, data);
      return;
    }
    if (data is String) {
      string(buffers, data);
      return;
    }
    if (data is num) {
      number(buffers, data);
      return;
    }
    if (data is List) {
      list(buffers, data);
      return;
    }
    if (data is bool) {
      number(buffers, data ? 1 : 0);
      return;
    }
    if (data is Map) {
      dict(buffers, data);
      return;
    }
    // bencode.js can access ArrayBufferView and ArrayBuffer, I ignore these type:
    // case 'arraybufferview': buffer(buffers, Buffer.from(data.buffer, data.byteOffset, data.byteLength)); break;
    // case 'arraybuffer': buffer(buffers, Buffer.from(data)); break;
  }

  void buffer(List<Uint8List> buffers, Uint8List data) {
    buffers.add(Uint8List.fromList(utf8.encode('${data.length}:')));
    buffers.add(data);
  }

  void string(List<Uint8List> buffers, String data) {
    var bytesLength = Uint8List.fromList(data.codeUnits).lengthInBytes;
    buffers.add(Uint8List.fromList('$bytesLength:$data'.codeUnits));
  }

  void number(List<Uint8List> buffers, num data) {
    buffers.add(Uint8List.fromList(utf8.encode('i${data}e')));
  }

  void dict(List<Uint8List> buffers, Map data) {
    buffers.add(buffD);

    var j = 0;
    dynamic k;
    var keys = data.keys.toList()..sort();
    var kl = keys.length;

    for (; j < kl; j++) {
      k = keys[j];
      if (data[k] == null) continue;
      string(buffers, k as String);
      _encode(buffers, data[k]);
    }

    buffers.add(buffE);
  }

  void list(List<Uint8List> buffers, List data) {
    var c = data.length;
    buffers.add(buffL);

    for (var i = 0; i < c; i++) {
      if (data[i] == null) continue;
      _encode(buffers, data[i]);
    }

    buffers.add(buffE);
  }
}

///
/// Decode bencoding format bytes to object.
///
/// This method comes from https://github.com/benjreinhart/bencode-js , just transfer JS to Dart
///
/// If don't provide string encoding , decoder won't decode the string field , just return ```Uint8List```
dynamic decode(Uint8List data, {int? start, int? end, String? stringEncoding}) {
  if (data.isEmpty) {
    return null;
  } else {
    return _Decode(data, start: start, end: end, stringEncoding: stringEncoding)
        .next();
  }
}

class _Decode {
  static const int INTEGER_START = 0x69; // 'i'
  static const int STRING_DELIM = 0x3A; // ':'
  static const int DICTIONARY_START = 0x64; // 'd'
  static const int LIST_START = 0x6C; // 'l'
  static const int END_OF_TYPE = 0x65; // 'e'

  int _position = 0;
  String? _stringEncoding;
  Uint8List _data;

  _Decode(this._data, {int? start, int? end, String? stringEncoding}) {
    _stringEncoding = stringEncoding?.toLowerCase();
    if (end != null && start != null) {
      _data = _data.sublist(start, end);
    }
  }

  int getIntFromBuffer(Uint8List buffer, int start, int end) {
    var sum = 0;
    var sign = 1;

    for (var i = start; i < end; i++) {
      var num = buffer[i];

      if (num < 58 && num >= 48) {
        sum = sum * 10 + (num - 48);
        continue;
      }

      if (i == start && num == 43) {
        // +
        continue;
      }

      if (i == start && num == 45) {
        // -
        sign = -1;
        continue;
      }

      if (num == 46) {
        // .
        // its a float. break here.
        break;
      }

      throw Exception('not a number: buffer[$i] = $num');
    }
    return sum * sign;
  }

  dynamic next() {
    if (_data.isEmpty) return null;
    switch (_data[_position]) {
      case DICTIONARY_START:
        return dictionary();
      case LIST_START:
        return list();
      case INTEGER_START:
        return integer();
      default:
        return buffer();
    }
  }

  int find(chr) {
    var i = _position;
    var c = _data.length;
    var d = _data;

    while (i < c) {
      if (d[i] == chr) return i;
      i++;
    }
    throw Exception(
        'Invalid data: Missing delimiter "${String.fromCharCode(chr)}" [0x${(chr as int).toRadixString(16)}]');
  }

  Map<String, dynamic> dictionary() {
    _position++;

    var dict = <String, dynamic>{};

    while (_data[_position] != END_OF_TYPE) {
      var keyBuffer = buffer();
      // 这里要注意，有时候返回的key字符串用utf-8解析是会出错的
      // 比如访问某infohash是利用latin1编码的，这里用utf-8解码就会出错
      // 但是，在规范中规定几乎都要使用utf8，所以这里做一个判断，能解就接，不能接就换
      if (keyBuffer is! String) {
        try {
          keyBuffer = utf8.decode(keyBuffer);
        } catch (e) {
          keyBuffer = String.fromCharCodes(keyBuffer);
        }
      }
      dict[keyBuffer] = next();
    }

    _position++;

    return dict;
  }

  List list() {
    _position++;

    var lst = [];

    while (_data[_position] != END_OF_TYPE) {
      lst.add(next());
    }

    _position++;

    return lst;
  }

  int integer() {
    var end = find(END_OF_TYPE);
    var number = getIntFromBuffer(_data, _position + 1, end);

    _position = end + 1;

    return number;
  }

  dynamic buffer() {
    var sep = find(STRING_DELIM);
    var length = getIntFromBuffer(_data, _position, sep);
    var end = ++sep + length;

    _position = end;
    var sublist = _data.sublist(sep, end);
    var encoder =
        _stringEncoding == null ? null : Encoding.getByName(_stringEncoding);
    return encoder != null ? encoder.decode(sublist) : sublist;
  }
}


lib/src/b_encode_decode_base.dart
import 'dart:convert';
import 'dart:typed_data';

///
/// Encode objects to bencoding format bytes;
///
/// This method comes from https://github.com/benjreinhart/bencode-js , just transfer JS to Dart
Uint8List encode(dynamic data, [dynamic buffer, int? offset]) {
  if (data == null) return Uint8List(0);
  return _Encode(data, buffer, offset).encoding();
}

class _Encode {
  int bytes = -1;
  final dynamic _data;
  final Uint8List? _buffer;
  final int? _offset;

  var buffE = Uint8List.fromList(utf8.encode('e')); // Buffer.from('e');
  var buffD = Uint8List.fromList(utf8.encode('d')); //Buffer.from('d');
  var buffL = Uint8List.fromList(utf8.encode('l')); //Buffer.from('l');

  _Encode(this._data, [this._buffer, this._offset]);

  Uint8List encoding() {
    var buffers = <Uint8List>[];
    Uint8List result;

    _encode(buffers, _data);
    result =
        Uint8List.fromList(buffers.fold<List<int>>([], (previousValue, buffer) {
      previousValue.addAll(buffer);
      return previousValue;
    }));
    bytes = result.length;

    if (_buffer != null) {
      (_buffer as Uint8List).insertAll(_offset ?? 0, result);
      return _buffer!;
    }

    return result;
  }

  void _encode(buffers, data) {
    if (data == null) {
      return;
    }

    if (data is Uint8List) {
      buffer(buffers, data);
      return;
    }
    if (data is String) {
      string(buffers, data);
      return;
    }
    if (data is num) {
      number(buffers, data);
      return;
    }
    if (data is List) {
      list(buffers, data);
      return;
    }
    if (data is bool) {
      number(buffers, data ? 1 : 0);
      return;
    }
    if (data is Map) {
      dict(buffers, data);
      return;
    }
    // bencode.js can access ArrayBufferView and ArrayBuffer, I ignore these type:
    // case 'arraybufferview': buffer(buffers, Buffer.from(data.buffer, data.byteOffset, data.byteLength)); break;
    // case 'arraybuffer': buffer(buffers, Buffer.from(data)); break;
  }

  void buffer(List<Uint8List> buffers, Uint8List data) {
    buffers.add(Uint8List.fromList(utf8.encode('${data.length}:')));
    buffers.add(data);
  }

  void string(List<Uint8List> buffers, String data) {
    var bytesLength = Uint8List.fromList(data.codeUnits).lengthInBytes;
    buffers.add(Uint8List.fromList('$bytesLength:$data'.codeUnits));
  }

  void number(List<Uint8List> buffers, num data) {
    buffers.add(Uint8List.fromList(utf8.encode('i${data}e')));
  }

  void dict(List<Uint8List> buffers, Map data) {
    buffers.add(buffD);

    var j = 0;
    dynamic k;
    var keys = data.keys.toList()..sort();
    var kl = keys.length;

    for (; j < kl; j++) {
      k = keys[j];
      if (data[k] == null) continue;
      string(buffers, k as String);
      _encode(buffers, data[k]);
    }

    buffers.add(buffE);
  }

  void list(List<Uint8List> buffers, List data) {
    var c = data.length;
    buffers.add(buffL);

    for (var i = 0; i < c; i++) {
      if (data[i] == null) continue;
      _encode(buffers, data[i]);
    }

    buffers.add(buffE);
  }
}

///
/// Decode bencoding format bytes to object.
///
/// This method comes from https://github.com/benjreinhart/bencode-js , just transfer JS to Dart
///
/// If don't provide string encoding , decoder won't decode the string field , just return ```Uint8List```
dynamic decode(Uint8List data, {int? start, int? end, String? stringEncoding}) {
  if (data.isEmpty) {
    return null;
  } else {
    return _Decode(data, start: start, end: end, stringEncoding: stringEncoding)
        .next();
  }
}

class _Decode {
  // ignore: constant_identifier_names
  static const int INTEGER_START = 0x69; // 'i'
  // ignore: constant_identifier_names
  static const int STRING_DELIM = 0x3A; // ':'
  // ignore: constant_identifier_names
  static const int DICTIONARY_START = 0x64; // 'd'
  // ignore: constant_identifier_names
  static const int LIST_START = 0x6C; // 'l'
  // ignore: constant_identifier_names
  static const int END_OF_TYPE = 0x65; // 'e'

  int _position = 0;
  String? _stringEncoding;
  Uint8List _data;

  _Decode(this._data, {int? start, int? end, String? stringEncoding}) {
    _stringEncoding = stringEncoding?.toLowerCase();
    if (end != null && start != null) {
      _data = _data.sublist(start, end);
    }
  }

  int getIntFromBuffer(Uint8List buffer, int start, int end) {
    var sum = 0;
    var sign = 1;

    for (var i = start; i < end; i++) {
      var num = buffer[i];

      if (num < 58 && num >= 48) {
        sum = sum * 10 + (num - 48);
        continue;
      }

      if (i == start && num == 43) {
        // +
        continue;
      }

      if (i == start && num == 45) {
        // -
        sign = -1;
        continue;
      }

      if (num == 46) {
        // .
        // its a float. break here.
        break;
      }

      throw Exception('not a number: buffer[$i] = $num');
    }
    return sum * sign;
  }

  dynamic next() {
    if (_data.isEmpty) return null;
    switch (_data[_position]) {
      case DICTIONARY_START:
        return dictionary();
      case LIST_START:
        return list();
      case INTEGER_START:
        return integer();
      default:
        return buffer();
    }
  }

  int find(chr) {
    var i = _position;
    var c = _data.length;
    var d = _data;

    while (i < c) {
      if (d[i] == chr) return i;
      i++;
    }
    throw Exception(
        'Invalid data: Missing delimiter "${String.fromCharCode(chr)}" [0x${(chr as int).toRadixString(16)}]');
  }

  Map<String, dynamic> dictionary() {
    _position++;

    var dict = <String, dynamic>{};

    while (_data[_position] != END_OF_TYPE) {
      var keyBuffer = buffer();
      // It should be noted here that sometimes the returned key string will be wrong when parsed with utf-8
      // For example, an infohash is encoded by latin1, and it will be wrong to decode it with utf-8.
      // However, it is stipulated in the specification that utf8 is almost always used
      if (keyBuffer is! String) {
        try {
          keyBuffer = utf8.decode(keyBuffer);
        } catch (e) {
          keyBuffer = String.fromCharCodes(keyBuffer);
        }
      }
      dict[keyBuffer] = next();
    }

    _position++;

    return dict;
  }

  List list() {
    _position++;

    var lst = [];

    while (_data[_position] != END_OF_TYPE) {
      lst.add(next());
    }

    _position++;

    return lst;
  }

  int integer() {
    var end = find(END_OF_TYPE);
    var number = getIntFromBuffer(_data, _position + 1, end);

    _position = end + 1;

    return number;
  }

  dynamic buffer() {
    var sep = find(STRING_DELIM);
    var length = getIntFromBuffer(_data, _position, sep);
    var end = ++sep + length;

    _position = end;
    var sublist = _data.sublist(sep, end);
    var encoder =
        _stringEncoding == null ? null : Encoding.getByName(_stringEncoding);
    return encoder != null ? encoder.decode(sublist) : sublist;
  }
}
